File: Python/FHIRSearch.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Create FHIR search params from NoSQL-like query structures.

try:
    from urllib import quote_plus
except Exception as e:
    from urllib.parse import quote_plus


class FHIRSearch(object):
    """ Create FHIR search params from NoSQL-like query structures.
    """
    
    def __init__(self, struct):
        if dict != type(struct):
            raise Exception("Must pass a Python dictionary, but got a {}".format(type(struct)))
        
        self.params = []
        for key, val in struct.items():
            self.params.append(FHIRSearchParam(key, val))
    
    def construct(self):
        parts = []
        if self.params is not None:
            for param in self.params:
                for expanded in param.expand():
                    parts.append(expanded.as_parameter())
        return '&'.join(parts)


class FHIRSearchParam(object):
    """ Holds one search parameter.
    """
    
    def __init__(self, name, value):
        self.name = name
        self.value = value
        self.handler = None
    
    def copy(self):
        clone = object.__new__(self.__class__)
        clone.__dict__ = self.__dict__.copy()
        return clone
    
    def expand(self):
        """ Parse the receiver's value and return a list of directly usable
        FHIRSearchParam instances.
        """
        if self.handler is None:
            self.handler = FHIRSearchParamHandler.handler_for(self.name)(None, self.value)
        self.handler.prepare()
        return self.handler.apply(self)
    
    def as_parameter(self):
        return '{}={}'.format(self.name, quote_plus(self.value, safe=',<=>'))


class FHIRSearchParamHandler(object):
    handles = None
    handlers = []
    
    @classmethod
    def announce_handler(cls, handler):
        cls.handlers.append(handler)
    
    @classmethod
    def handler_for(cls, key):
        for handler in cls.handlers:
            if handler.can_handle(key):
                return handler
        return cls
    
    @classmethod
    def can_handle(cls, key):
        if cls.handles is not None:
            return key in cls.handles
        return True         # base class handles everything else, so be sure to test it last!
    
    
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.modifier = []
        self.multiplier = []
    
    def prepare(self, parent=None):
        if dict == type(self.value):
            for key, val in self.value.items():
                handler = FHIRSearchParamHandler.handler_for(key)(key, val)
                handler.prepare(self)
        
        if parent is not None:
            parent.multiplier.append(self)
    
    def apply(self, param):
        for handler in self.modifier:
            handler.apply(param)
        
        self._apply(param)
        
        # if we have multiplier, expand sequentially
        if len(self.multiplier) > 0:
            expanded = []
            for handler in self.multiplier:
                clone = param.copy()
                expanded.extend(handler.apply(clone))
            
            return expanded
        
        # no multiplier, just return the passed-in paramater
        return [param]
    
    def _apply(self, param):
        if self.key is not None:
            param.name = '{}.{}'.format(param.name, self.key)
        if 0 == len(self.multiplier):
            param.value = self.value


class FHIRSearchParamModifierHandler(FHIRSearchParamHandler):
    modifiers = {
        '$asc': ':asc',
        '$desc': ':desc',
        '$exact': ':exact',
        '$missing': ':missing',
        '$null': ':missing',
        '$text': ':text',
    }
    handles = modifiers.keys()
    
    def _apply(self, param):
        if self.key not in self.__class__.modifiers:
            raise Exception('Unknown modifier "{}" for "{}"'.format(self.key, param.name))
        param.name += self.__class__.modifiers[self.key]
        param.value = self.value


class FHIRSearchParamOperatorHandler(FHIRSearchParamHandler):
    operators = {
        '$gt': '>',
        '$lt': '<',
        '$lte': '<=',
        '$gte': '>=',
    }
    handles = operators.keys()
    
    def _apply(self, param):
        if self.key not in self.__class__.operators:
            raise Exception('Unknown operator "{}" for "{}"'.format(self.key, parent.name))
        param.value = self.__class__.operators[self.key] + self.value


class FHIRSearchParamMultiHandler(FHIRSearchParamHandler):
    handles = ['$and', '$or']
    
    def prepare(self, parent):
        if list != type(self.value):
            raise Exception('Expecting a list argument for "{}" but got {}'.format(parent.key, self.value))
        
        handlers = []
        for val in self.value:
            if dict == type(val):
                for kkey, vval in val.items():
                    handlers.append(FHIRSearchParamHandler.handler_for(kkey)(kkey, vval))
            else:
                handlers.append(FHIRSearchParamHandler.handler_for(parent.key)(None, val))
        
        if '$and' == self.key:
            for handler in handlers:
                handler.prepare(parent)
        elif '$or' == self.key:
            ors = [h.value for h in handlers]
            handler = FHIRSearchParamHandler.handler_for(parent.key)(None, ','.join(ors))
            handler.prepare(parent)
        else:
            raise Exception('I cannot handle "{}"'.format(self.key))


class FHIRSearchParamTypeHandler(FHIRSearchParamHandler):
    handles = ['$type']
    
    def prepare(self, parent):
        parent.modifier.append(self)
    
    def _apply(self, param):
        param.name = '{}:{}'.format(param.name, self.value)
    

# announce all handlers
FHIRSearchParamHandler.announce_handler(FHIRSearchParamModifierHandler)
FHIRSearchParamHandler.announce_handler(FHIRSearchParamOperatorHandler)
FHIRSearchParamHandler.announce_handler(FHIRSearchParamMultiHandler)
FHIRSearchParamHandler.announce_handler(FHIRSearchParamTypeHandler)


if '__main__' == __name__:
    print('1 '+FHIRSearch({'name': 'Willis'}).construct())
    print('1 name=Willis')
    print('')
    print('2 '+FHIRSearch({'name': {'$exact': 'Willis'}}).construct())
    print('2 name:exact=Willis')
    print('')
    print('3 '+FHIRSearch({'name': {'$or': ['Willis', 'Wayne', 'Bruce']}}).construct())
    print('3 name=Willis,Wayne,Bruce')
    print('')
    print('4 '+FHIRSearch({'name': {'$and': ['Willis', {'$exact': 'Bruce'}]}}).construct())
    print('4 name=Willis&name:exact=Bruce')
    print('')
    print('5 '+FHIRSearch({'birthDate': {'$gt': '1950', '$lte': '1970'}}).construct())
    print('5 birthDate=>1950&birthDate=<=1970')
    print('')
    print('6 '+FHIRSearch({'subject.name': {'$exact': 'Willis'}}).construct())
    print('6 subject.name:exact=Willis')
    print('')
    print('7 '+FHIRSearch({'subject': {'$type': 'Patient', 'name': 'Willis', 'birthDate': {'$gt': '1950', '$lte': '1970'}}}).construct())
    print('7 subject:Patient.name=Willis&subject:Patient.birthDate=>1950&subject:Patient.birthDate=<=1970')
    

File: Swift/FHIRSearch.swift
//
//  FHIRSearch.swift
//  SMART-on-FHIR
//
//  Created by Pascal Pfiffner on 7/10/14.
//  2014, SMART Platforms.
//

import Foundation

let FHIRSearchErrorDomain = "FHIRSearchErrorDomain"


/**
	Instances of this class can perform searches on a server.

	Searches are instantiated from MongoDB-like query constructs, like:

	    let srch = Patient.search(["address": "Boston", "gender": "male", "given": ["$exact": "Willis"]])

	Then srch.perform() will run the following URL query against the server:

	    "Patient?address=Boston&gender=male&given:exact=Willis"
 */
public class FHIRSearch
{
	/// The first search parameter must define a profile type to which the search is applied.
	public var profileType: FHIRResource.Type?
	
	/// The query construct used to describe the search
	let query: FHIRSearchConstruct
	
	/// The sorting to request. Use tuples with the value followed by "asc" or "desc": [("given", "asc"), ("family", "asc")].
	public var sort: [(String, String)]?
	
	/// The number of results to return per page; leave nil to let the server decide.
	public var pageCount: Int?
	
	/// The URL to retrieve the next page of results from; nil if there are no more results.
	var nextPageURL: NSURL?
	
	var busy = false
	
	/// Returns true if there are more search results to be fetched.
	public var hasMore: Bool {
		return (nil != nextPageURL)
	}
	
	
	/** Designated initializer. */
	public init(query: AnyObject) {
		self.query = FHIRSearchConstruct(construct: query)
	}
	
	/** Convenience initializer. */
	public convenience init(type: FHIRResource.Type, query: AnyObject) {
		self.init(query: query)
		profileType = type
	}
	
	
	// MARK: - Running Search
	
	func reset() {
		nextPageURL = nil
	}
	
	/**
		Creates the relative server path and query URL string.
	 */
	public func construct() -> String {
		var extra = [FHIRURLParam]()
		if let count = pageCount {
			extra.append(FHIRURLParam(name: "_count", value: "\(count)"))
		}
		if let sorters = sort {
			for (val, ord) in sorters {
				extra.append(FHIRURLParam(name: "_sort:\(ord)", value: val))
			}
		}
		
		// expand
		let qry = query.expand(extraArguments: extra)
		if let type = profileType {
			if countElements(qry) > 0 {
				return "\(type.resourceName)?\(qry)"
			}
			return type.resourceName
		}
		if countElements(qry) > 0 {
			return "?\(qry)"
		}
		return ""
	}
	
	/**
		Performs a GET on the server after constructing the query URL, returning an error or a bundle resource with the
		callback.
	
		Calling this method will always restart search, not fetch subsequent pages.
	
		:param: server The FHIRServer instance on which to perform the search
		:param: callback The callback, receives the response Bundle or an NSError message describing what went wrong
	 */
	public func perform(server: FHIRServer, callback: ((bundle: Bundle?, error: NSError?) -> Void)) {
		if nil == profileType {
			let err = NSError(domain: FHIRSearchErrorDomain, code: 0, userInfo: [NSLocalizedDescriptionKey: "Cannot find the profile type against which to run the search"])
			callback(bundle: nil, error: err)
			return
		}
		
		reset()
		performSearch(server, queryPath: construct(), callback: callback)
	}
	
	/**
		Attempts to retrieve the next page of search results. If there are none, the callback is called immediately
		with no bundle and no error.
	
		:param: server The FHIRServer instance on which to perform the search
		:param: callback The callback, receives the response Bundle or an NSError message describing what went wrong
	 */
	public func nextPage(server: FHIRServer, callback: ((bundle: Bundle?, error: NSError?) -> Void)) {
		if let next = nextPageURL?.absoluteString {
			performSearch(server, queryPath: next, callback: callback)
		}
		else {
			callback(bundle: nil, error: nil)
		}
	}
	
	func performSearch(server: FHIRServer, queryPath: String, callback: ((bundle: Bundle?, error: NSError?) -> Void)) {
		if busy {
			callback(bundle: nil, error: nil)
			return
		}
		
		busy = true
		server.getJSON(queryPath) { response in
			self.busy = false
			
			if let error = response.error {
				callback(bundle: nil, error: error)
			}
			else {
				let bundle = Bundle(json: response.json)
				bundle._server = server
				if let entries = bundle.entry {
					for entry in entries {
						entry.resource?._server = server		// workaround for when "Bundle" gets deallocated
					}
				}
				
				// is there more?
				self.nextPageURL = nil
				if let links = bundle.link {
					for link in links {
						if "next" == link.relation {
							self.nextPageURL = link.url
							break
						}
					}
				}
				
				callback(bundle: bundle, error: nil)
			}
		}
	}
}


/**
	Instances of this struct represent URL query parameters.
 */
struct FHIRURLParam
{
	/// Parameter name
	let name: String
	
	/// Parameter value
	let value: String
	
	
	func asParameter() -> String {
		return "\(name)=\(value)"		// TODO: encode value
	}
}


/**
	This class is used to create FHIRURLParam instances from FHIRSearchConstruct objects.
 */
class FHIRSearchParam: Printable
{
	var name: String?
	var isModifier = false
	var value: String?
	weak var parent: FHIRSearchParam?
	var children: [FHIRSearchParam]? {
		didSet {
			if let chldrn = children {
				for child in chldrn {
					child.parent = self
				}
			}
		}
	}
	var description: String {
		return "<FHIRSearchParam> \(name ?? nil) [parent \(parent?.description ?? nil) and \(nil != children ? countElements(children!) : 0) children]"
	}
	
	init(name: String, parent: FHIRSearchParam?) {
		self.name = name
		self.parent = parent
	}
	
	init(value: String, parent: FHIRSearchParam?) {
		self.value = value
		self.parent = parent
	}
	
	/** Instantiate from any object, delegating to FHIRSearchConstruct to figure out what the object means. */
	class func from(any: AnyObject, parent: FHIRSearchParam?) -> [FHIRSearchParam] {
		if let str = any as? String {
			return [FHIRSearchParam(value: str, parent: parent)]
		}
		if let bol = any as? Bool {
			return [FHIRSearchParam(value: bol ? "true" : "false", parent: parent)]
		}
		
		let construct = FHIRSearchConstruct(construct: any)
		return construct.prepare(parent)
	}
	
	/** Recursively determine the parameter name, looking at all parent objects. */
	func parentName() -> String? {
		var full = name
		if let prnt = parent?.parentName() {
			if nil != full {
				if isModifier {
					full = "\(prnt)\(full!)"
				}
				else {
					full = "\(prnt).\(full!)"
				}
			}
			else {
				full = prnt
			}
		}
		return full
	}
	
	/** Expand all children to instantiate FHIRURLParam objects. */
	func expand() -> [FHIRURLParam] {
		if let chldren = children {
			var arr = [FHIRURLParam]()
			for child in chldren {
				arr.extend(child.expand())
			}
			return arr
		}
		else if let val = value {
			return [FHIRURLParam(name: parentName() ?? "", value: val)]
		}
		return []
	}
}


struct FHIRSearchConstruct
{
	static var handlers: [FHIRSearchConstructHandler] = [
		FHIRSearchConstructAndHandler(),
		FHIRSearchConstructOrHandler(),
		FHIRSearchConstructModifierHandler(),
		FHIRSearchConstructOperatorHandler(),
		FHIRSearchConstructTypeHandler(),
	]
	
	static func handlerFor(key: String) -> FHIRSearchConstructHandler? {
		for handler in self.handlers {
			if handler.handles(key) {
				return handler
			}
		}
		return nil
	}
	
	
	let construct: AnyObject
	
	init(construct: AnyObject) {
		self.construct = construct
	}
	
	func expand(extraArguments: [FHIRURLParam]? = nil) -> String {
		var arr = [String]()
		for params in self.prepare(nil) {
			for param in params.expand() {
				arr.append(param.asParameter())
			}
		}
		if let extras = extraArguments {
			for extra in extras {
				arr.append(extra.asParameter())
			}
		}
		
		return "&".join(arr)
	}
	
	func prepare(parent: FHIRSearchParam?) -> [FHIRSearchParam] {
		var arr = [FHIRSearchParam]()
		if let myarr = construct as? [AnyObject] {
			for any in myarr {
				let sub = FHIRSearchConstruct(construct: any)
				arr.extend(sub.prepare(nil))
			}
			return arr
		}
		
		if let dict = construct as? [String: AnyObject] {
			for (key, val) in dict {
				//println("-> \(key): \(val)")
				var param = FHIRSearchParam(name: key, parent: parent)
				
				// special handling?
				if let handler = self.dynamicType.handlerFor(key) {
					handler.handle(param, value: val)
				}
					
				// this is a sub-structure, expand
				else if let dict = val as? [String: AnyObject] {
					let construct = FHIRSearchConstruct(construct: dict)
					param.children = construct.prepare(param)
				}
					
				// a string
				else if let str = val as? String {
					param.value = str
				}
				else {
					println("WARNING: no idea what to do with \(key): \(val), ignoring")
				}
				arr.append(param)
			}
			return arr
		}
		
		println("WARNING: not sure what to do with \"\(construct)\"")
		return arr
	}
}


// MARK: - Special Handlers

protocol FHIRSearchConstructHandler
{
	func handles(key: String) -> Bool
	func handle(param: FHIRSearchParam, value: AnyObject)
}


struct FHIRSearchConstructAndHandler: FHIRSearchConstructHandler
{
	func handles(key: String) -> Bool {
		return ("$and" == key)
	}
	
	func handle(param: FHIRSearchParam, value: AnyObject) {
		if let arr = value as? [AnyObject] {
			param.name = nil
			var ret = [FHIRSearchParam]()
			for obj in arr {
				ret.extend(FHIRSearchParam.from(obj, parent: param.parent))
			}
			
			if nil != param.children {
				param.children!.extend(ret)
			}
			else {
				param.children = ret
			}
		}
		else {
			println("ERROR: must supply an array of objects to an $and modifier")
		}
	}
}


struct FHIRSearchConstructOrHandler: FHIRSearchConstructHandler
{
	func handles(key: String) -> Bool {
		return ("$or" == key)
	}
	
	func handle(param: FHIRSearchParam, value: AnyObject) {
		if let arr = value as? [AnyObject] {
			var strs = [String]()
			for obj in arr {
				if let str = obj as? String {
					strs.append(str)
				}
				else {
					println("WARNING: what do I do with \(obj)?")
				}
			}
			param.name = nil
			param.value = ",".join(strs)
		}
		else {
			println("ERROR: must supply an array of objects to an $or modifier")
		}
	}
}


struct FHIRSearchConstructModifierHandler: FHIRSearchConstructHandler
{
	static let map = [
//		"$asc": ":asc",
//		"$desc": ":desc",
		"$exact": ":exact",
		"$missing": ":missing",
		"$null": ":missing",
		"$text": ":text",
	]
	
	func handles(key: String) -> Bool {
		return contains(FHIRSearchConstructModifierHandler.map.keys, key)
	}
	
	func handle(param: FHIRSearchParam, value: AnyObject) {
		if let modifier = FHIRSearchConstructModifierHandler.map[param.name ?? ""] {
			param.name = modifier
			param.isModifier = true
			param.children = FHIRSearchParam.from(value, parent: param)
		}
		else {
			println("ERROR: unknown modifier \(param.name)")
		}
	}
}


struct FHIRSearchConstructOperatorHandler: FHIRSearchConstructHandler
{
	static let map = [
		"$gt": "%3E",           // NSURL() fails if un-encoded ">" and "<" are present in the query part
		"$lt": "%3C",
		"$lte": "%3C%3D",       // NSURL() does not fail on "=" but let's also encode these to be consistent
		"$gte": "%3E%3D",
	]
	
	func handles(key: String) -> Bool {
		return contains(FHIRSearchConstructOperatorHandler.map.keys, key)
	}
	
	func handle(param: FHIRSearchParam, value: AnyObject) {
		if let modifier = FHIRSearchConstructOperatorHandler.map[param.name!] {
			if let str = value as? String {
				param.name = nil
				param.value = "\(modifier)\(str)"
			}
		}
		else {
			println("ERROR: unknown operator \(param.name) for \(value)")
		}
	}
}


struct FHIRSearchConstructTypeHandler: FHIRSearchConstructHandler
{
	func handles(key: String) -> Bool {
		return ("$type" == key)
	}
	
	func handle(param: FHIRSearchParam, value: AnyObject) {
		if let type = value as? String {
			if let parent = param.parent {
				parent.name = (parent.name ?? "") + ":\(type)"
			}
			else {
				println("ERROR: must have a parent parameter to use $type")
			}
		}
		else {
			println("ERROR: must supply a String to a $type modifier, got \(value)")
		}
	}
}

